### ISSUE TITLE
Dataset.FlatIterator() leaks feeding goroutine if not exhausted.

### ISSUE THREAD

**Issue by bpeake-illuscio** [suyashkumar/dicom#183]
Hello!

While writing some code that uses Dataset.FlatIterator(), I realized that unless the Iterator is exhausted, a goroutine is leaked. That is, the goroutine feeding the iterator will never exit.

Here's a compact example:

```go
package main

import (
	"errors"
	"fmt"
	"github.com/suyashkumar/dicom"
	"log"
	"runtime"
)

var DicomPaths = []string{
	"./testdata/1.dcm",
	"./testdata/2.dcm",
	"./testdata/3.dcm",
	"./testdata/4.dcm",
	"./testdata/5.dcm",
}

// DoSomething will be used to simulate an error occurring during iteration.
func DoSomething(element *dicom.Element) error {
	return errors.New("error doing something with element")
}

// DatasetInspector ranges ofer Dataset.FlatIterator() and does something that can fail
// with each element.
func DatasetInspector(filepath string) error {
	// First we read the dataset
	dataset, err := dicom.ParseFile(filepath, nil)
	if err != nil {
		panic(err)
	}

	// Now we iterate over all of our elements and call DoSomething() on each.
	for element := range dataset.FlatIterator() {
		err = DoSomething(element)
		if err != nil {
			// THIS ERROR RETURN WILL LEAK THE GOROUTINE FEEDING OUR RANGE LOOP
			return err
		}
	}

	return nil
}

func main() {
	// Call our inspector on 5 dicom files. Each one should error and will result in a leaked goroutine.
	for _, thisFile := range DicomPaths {
		err := DatasetInspector(thisFile)
		if err != nil {
			log.Println("ERROR:", err)
		}
	}

	// Force garbage collection to show these goroutines are not cleaned up.
	runtime.GC()

	// If we print the runtime stack we'll see that we have leaked goroutines.
	buffer := make([]byte, 5012)
	runtime.Stack(buffer, true)
	fmt.Println("\nCURRENT STACK:\n", string(buffer))
}
```

Now let's look at our output:

```
2021/03/10 13:32:10 ERROR: error doing something with element
2021/03/10 13:32:10 ERROR: error doing something with element
2021/03/10 13:32:10 ERROR: error doing something with element
2021/03/10 13:32:10 ERROR: error doing something with element
2021/03/10 13:32:10 ERROR: error doing something with element

CURRENT STACK:
 goroutine 1 [running]:
main.main()
        .../dicom/example/main.go:60 +0xf6

goroutine 6 [chan send]:
github.com/suyashkumar/dicom.flatElementsIterator(0xc000106400, 0x6d, 0x80, 0xc0000200c0)
        .../dicom/dataset.go:85 +0x147
github.com/suyashkumar/dicom.(*Dataset).FlatIterator.func1(0xc00000c0d8, 0xc0000200c0)
        .../dicom/dataset.go:70 +0x4a
created by github.com/suyashkumar/dicom.(*Dataset).FlatIterator
        .../dicom/dataset.go:69 +0x67

goroutine 7 [chan send]:
github.com/suyashkumar/dicom.flatElementsIterator(0xc000106800, 0x6c, 0x80, 0xc000020120)
        .../dicom/dataset.go:85 +0x147
github.com/suyashkumar/dicom.(*Dataset).FlatIterator.func1(0xc00000d8c0, 0xc000020120)
        .../dicom/dataset.go:70 +0x4a
created by github.com/suyashkumar/dicom.(*Dataset).FlatIterator
        .../dicom/dataset.go:69 +0x67

goroutine 37 [chan send]:
github.com/suyashkumar/dicom.flatElementsIterator(0xc000108800, 0x86, 0x100, 0xc000020180)
        .../dicom/dataset.go:85 +0x147
github.com/suyashkumar/dicom.(*Dataset).FlatIterator.func1(0xc0002b4ff0, 0xc000020180)
        .../dicom/dataset.go:70 +0x4a
created by github.com/suyashkumar/dicom.(*Dataset).FlatIterator
        .../dicom/dataset.go:69 +0x67

goroutine 13 [chan send]:
github.com/suyashkumar/dicom.flatElementsIterator(0xc000106000, 0x59, 0x80, 0xc0000201e0)
        .../dicom/dataset.go:85 +0x147
github.com/suyashkumar/dicom.(*Dataset).FlatIterator.func1(0xc00000c0d8, 0xc0000201e0)
        .../dicom/dataset.go:70 +0x4a
created by github.com/suyashkumar/dicom.(*Dataset).FlatIterator
        .../dicom/dataset.go:69 +0x67

goroutine 14 [chan send]:
github.com/suyashkumar/dicom.flatElementsIterator(0xc000106c00, 0x52, 0x80, 0xc000020240)
        .../dicom/dataset.go:85 +0x147
github.com/suyashkumar/dicom.(*Dataset).FlatIterator.func1(0xc00009e198, 0xc000020240)
        .../dicom/dataset.go:70 +0x4a
created by github.com/suyashkumar/dicom.(*Dataset).FlatIterator
        .../dicom/dataset.go:69 +0x67


Process finished with exit code 0
```

You can see that even after garbage collection, all of the routines feeding our aborted iterations are still running. This means in a long-running service setting, any failure within the range loop will leak the goroutine feeding the iterator, which means also leaking the entire DICOM []*dicom.Element feeding it, as we can see from Goland's Debugger, with a breakpoint after our main loop:

<img width="1105" alt="Screen Shot 2021-03-10 at 1 44 16 PM" src="https://user-images.githubusercontent.com/44302546/110702063-e1eea600-81a6-11eb-8287-9ab881983a70.png">

The most straightforward fix, in my opinion, would be to make the method context-aware, so the code could be set up as follows:

```go
iterCtx, iterCancel := context.WithCancel(context.Background())
defer iterCancel()

// Now we iterate over all of our elements, looking for PN values.
for element := range dataset.FlatIterator(iterCtx) {
	// some code
}
```

But that is also a breaking change. I'm happy to take a crack at a PR if you would like, with this approach or another.

**Comment by suyashkumar** [comment]
Hi! Thanks for raising the issue and for posting a great minimal example. I agree that this is something we should address, particularly since that Go routine holds a reference to the read dataset (so the dataset presumably isn't cleaned up during GC even when it's not needed).

Semantically, I think using a context is reasonable though it's easy for the caller to mess up which I'm not a big fan of (e.g. if the caller just passes context.Background()).

While I like the channel semantics here, another option is to implement a stateful iterator object that doesn't run inside a goroutine (though it'll be a bit more work, and we'd have to think about what happens if the dataset object changes under the hood in another Go routine, though that can be explicitly out of scope). When that iterator object doesn't have any more references, it'll be destroyed as usual. 

By stateful iterator I mean something like
```
iter := dicom.FlatIterator(dataset)
for iter.HasNext() {
    val := iter.NextValue()
}
```

I could envision a pretty straightforward implementation, even if the dataset ends up mutating in the middle of the FlatIterator (though if we want we can call that out). 

I think I will prototype a stateful iterator approach just to see, as I think it would be more difficult for the caller to accidentally use incorrectly vs the context approach. Let me know if you have any thoughts on this!

**Comment by bpeake-illuscio** [comment]
I think your first instinct of moving to a stateful iterator is the better approach, personally. Implementing a Closer adds to the amount of bookkeeping the caller needs to keep track of, and will likely be confusing to a caller as we aren't really connecting to any sort of external resource. 

Whereas with a stateful iterator you can just let it go out of scope and the garbage collector will do its ting as normal.

**Comment by bpeake-illuscio** [comment]
Theres also a third option:

Rather than implementing this as an iterator, we could just return a flattened slice of elements, so the signature is something
more akin to:

```go
func (d *Dataset) Flattened() []*Element {
    // implementation
}
```
Where all the elements are pre-collected for you. Because no io is involved, it's not clear to me that using a more complicated iterator approach will be so much more performant that it will really make a difference.

**Comment by suyashkumar** [comment]
I considered this as well, however one consideration is that this gives the caller a static flattened copy of element pointers. It's possible if the underlying data set changes that this flattened copy is out of sync. Luckily if any of the individual elements change its fine (as we're holding pointers). Perhaps it should be common sense for the caller to watch out for situations like this.

in any case, a stateful iterator seems to be the most natural approach--doesn't suffer from the problem above, it's easily garbage collected, though the API is slightly more complex (e.g. can't just range over the return value sadly).

To unblock, perhaps I can go ahead and implement a simple stateful iterator that builds the Flattened() set of element pointers and then returns elements over it (as opposed to a possibly more complex implementation). If we wanted, we could consider exposing the underlying Flatten() operation as well with a note that it is a static snapshot of the Dataset. The API could remain the same, but there are several future optimizations we could do--for example, don't recompute the flattened dataset if the underlying dataset hasn't changed (e.g. cache it), etc. But API wise, getting folks to use the iterator seems reasonable to me. Thoughts?

**Comment by suyashkumar** [comment]
I've put up a draft in #184. I've added `FlatStatefulIterator` and added an ExhaustElementChannel that can be used to exhaust the channel in a defer for the channel-based API. I think we should consider deprecating the channel API, but it is in use today, and some consumer depend on the channel (e.g. having multiple processing goroutines read elements off of the channel). Likely would want to evaluate those use cases and deprecate the API release after next to ease the transition off of it.

**Comment by suyashkumar** [comment]
At some point either we should deprecate the channel-based API, or we should rename it so that it's not seen as the default for iteration (e.g. name it FlatChannelIterator or something). For most usecases, the stateful iterator should be sufficient, but having the channel approach around may still be useful for some users.

**Comment by bpeake-illuscio** [comment]
I think it makes sense to deprecate the channel version over time. I feel like concurrency-awareness should probably be left up to the caller, and be considered out-of-scope for the lib. Callers can implement distributing the results of the iterator over a channel.

**Comment by bpeake-illuscio** [comment]
Also let me know if I can help in anyway! (code review, et)

**Comment by suyashkumar** [comment]
yeah, as said before I think deprecation makes sense, though I think it would be nice to phase it out over the next 2 releases as mentioned above. I may also add some benchmarks to do some comparisons, but copying a bunch of pointers shouldn't be a big deal.

Feel free to take a look at the PR if you have any thoughts, my plan is to take another look late tonight or tomorrow before merging!

**Comment by suyashkumar** [comment]
Also, thank you again for raising this! Appreciate your contributions and feedback on the library!

**Comment by bpeake-illuscio** [comment]
No problem, it's a great lib and I'm happy that I can contribute to it!
